diffusion/models/layers.py

```diff
@@ -222,3 +223,208 @@ def __call__(
+    """
+
+    def __init__(self, input_channels: int, dropout_probability: float = 0.0):
+        super().__init__()
+        self.input_channels = input_channels
+        self.dropout_probability = dropout_probability
+        # Normalization layer. Here we're using groupnorm to be consistent with the original implementation.
+        self.norm = nn.GroupNorm(num_groups=32, num_channels=self.input_channels, eps=1e-6, affine=True)
+        # Conv layer to transform the input into q, k, and v
+        self.qkv_conv = nn.Conv2d(self.input_channels, 3 * self.input_channels, kernel_size=1, stride=1, padding=0)
+        # Init the qkv conv weights
+        nn.init.kaiming_normal_(self.qkv_conv.weight, nonlinearity='linear')
+        # Conv layer to project to the output.
+        self.proj_conv = nn.Conv2d(self.input_channels, self.input_channels, kernel_size=1, stride=1, padding=0)
+        nn.init.kaiming_normal_(self.proj_conv.weight, nonlinearity='linear')
+
+    def _reshape_for_attention(self, x: torch.Tensor) -> torch.Tensor:
+        """Reshape the input tensor for attention."""
+        # x is (B, C, H, W), need it to be (B, H*W, C) for attention
+        x = x.permute(0, 2, 3, 1).reshape(x.shape[0], -1, x.shape[1]).contiguous()
+        return x
+
+    def _reshape_from_attention(self, x: torch.Tensor, H: int, W: int) -> torch.Tensor:
+        """Reshape the input tensor from attention."""
+        # x is (B, H*W, C), need it to be (B, C, H, W) for conv
+        x = x.reshape(x.shape[0], H, W, -1).permute(0, 3, 1, 2).contiguous()
+        return x
+
+    def forward(self, x: torch.Tensor) -> torch.Tensor:
+        """Forward through the attention layer."""
+        # Need to remember H, W to get back to it
+        H, W = x.shape[2:]
+        h = self.norm(x)
+        # Get q, k, and v
+        qkv = self.qkv_conv(h)
+        q, k, v = torch.split(qkv, self.input_channels, dim=1)
+        # Use torch's built in attention function
+        q, k, v = self._reshape_for_attention(q), self._reshape_for_attention(k), self._reshape_for_attention(v)
+        h = F.scaled_dot_product_attention(q, k, v, dropout_p=self.dropout_probability)
+        # Reshape back into an image style tensor
+        h = self._reshape_from_attention(h, H, W)
+        # Project to the output
+        h = self.proj_conv(h)
+        return h
+
+
+class Downsample(nn.Module):
+    """Downsampling layer that downsamples by a factor of 2.
+
+    Args:
+        input_channels (int): Number of input channels.
+        resample_with_conv (bool): Whether to use a conv for downsampling.
+    """
+
+    def __init__(self, input_channels: int, resample_with_conv: bool):
+        super().__init__()
+        self.input_channels = input_channels
+        self.resample_with_conv = resample_with_conv
+        if self.resample_with_conv:
+            self.conv = nn.Conv2d(self.input_channels, self.input_channels, kernel_size=3, stride=2, padding=0)
```

Comment:
Is there a reason this kernel has to be 3 instead of 2 or 4?

Reply:
It's 3 from the original code [here](https://github.com/CompVis/stable-diffusion/blob/21f890f9da3cfbeaba8e2ac3c425ee9e998d5229/ldm/modules/diffusionmodules/model.py#L68). Are you asking about using 2 or 4 to avoid the weird padding?

Reply:
Yes haha, but if that is how they do it 🤷